### PR TITLE
[Merged by Bors] - Add thread create/destroy callbacks to TaskPool

### LINKED ISSUE #1991: Thread creation/destruction callbacks for TaskPool
## What problem does this solve or what need does it fill?

I'd like to be able to run something on each created thread in a TaskPool after it's spawned and before it's destroyed. This would be useful for instrumentation purposes, profilers.

## What solution would you like?

I propose adding two new boxed function fields in `TaskPoolBuilder` to define these operations.

```rs
TaskPoolBuilder::new()
    .num_threads(N)
    .thread_name("MyWorker")
    .on_thread_spawn(|ix| { /* executes during build() on each worker */ })
    .on_thread_destroy(|ix| { /* ... */ })
    .build()
```

## What alternative(s) have you considered?

I considered spawning dummy tasks right after creating a TaskPool, but I don't expect to get a stable, round-robin order with each spawned task.

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Fix #1991. Allow users to have a bit more control over the creation and finalization of the threads in `TaskPool`.

## Solution
Add new methods to `TaskPoolBuilder` that expose callbacks that are called to initialize and finalize each thread in the `TaskPool`.

Unlike the proposed solution in #1991, the callback is argument-less. If an an identifier is needed, `std::thread::current` should provide that information easily.

Added a unit test to ensure that they're being called correctly.

### CODE REVIEW

File: crates/bevy_tasks/src/task_pool.rs [hunk 1]
```diff
@@ -451,6 +484,46 @@ mod tests {
...
+        {
+            // Build and immediately drop to terminate
+            TaskPoolBuilder::new()
+                .num_threads(10)
+                .on_thread_spawn(move || {
+                    start_counter.fetch_add(1, Ordering::Relaxed);
+                })
+                .build();
+        }
+        assert_eq!(10, counter.load(Ordering::Relaxed));
```
Comment (hymm) [inline crates/bevy_tasks/src/task_pool.rs:500]: would it be possible to move this assert inside the scope with a barrier?

File: crates/bevy_tasks/src/task_pool.rs [hunk 2]
```diff
@@ -451,6 +484,46 @@ mod tests {
...
+        {
+            TaskPoolBuilder::new()
+                .num_threads(5)
+                .on_thread_spawn(move || {
+                    start_counter.fetch_add(1, Ordering::Relaxed);
+                })
+                .on_thread_destroy(move || {
+                    end_counter.fetch_sub(1, Ordering::Relaxed);
+                })
+                .build();
```
Comment (hymm) [inline crates/bevy_tasks/src/task_pool.rs:522]: if it is possible could we add an assert here

**Comment by hymm** [comment]
made a some minor comments on the test. Would be willing to approve without any changes

File: crates/bevy_tasks/src/task_pool.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_tasks/src/task_pool.rs:500]: The intent is to test how both initialization and finalization works. I'd rather keep it outside of the scope.

File: crates/bevy_tasks/src/task_pool.rs (hunk 2 above)
Comment (james7132) [reply crates/bevy_tasks/src/task_pool.rs:522]: See other comment.

Comment (hymm) [reply crates/bevy_tasks/src/task_pool.rs:522]: I'm not sure if I'm misunderstanding, but I was saying to do something like this to test that the on_thread_spawn was being called before the threads were destroyed.

```rust
let start_counter = counter.clone();
let end_counter = counter.clone();
let barrier = Arc::new(Barrier::new(6));
let last_barrier = barrier.clone();
{
    let _pool = TaskPoolBuilder::new()
        .num_threads(5)
        .on_thread_spawn(move || {
            start_counter.fetch_add(1, Ordering::Relaxed);
            barrier.clone().wait();
        })
        .on_thread_destroy(move || {
            end_counter.fetch_sub(1, Ordering::Relaxed);
        })
        .build();

    last_barrier.wait();
    assert_eq!(-5, counter.load(Ordering::Relaxed));
}
assert_eq!(-10, counter.load(Ordering::Relaxed));
```

**Review by hymm — APPROVED** [review]

File: crates/bevy_tasks/src/task_pool.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_tasks/src/task_pool.rs:505]: Done.

File: crates/bevy_tasks/src/task_pool.rs (hunk 2 above)
Comment (james7132) [reply crates/bevy_tasks/src/task_pool.rs:531]: Done.

**Review by mockersf — APPROVED** [review]

**Comment by james7132** [comment]
bors retry

**Pull request closed without merging**